### DISCUSSION TITLE
[Chapter4.1-4.5] 트랜지스터에서 CPU로, 이보다 더 중요한 것은 없다.

### DISCUSSION THREAD

**Discussion by thgr8ganzi** [elegant-functional-2023/secrets-of-computer#11] [General]
### 감상평
* 여러분의 생각을 들려주세요
* ~7/20 까지

**Comment by ella-yschoi** [comment]
## 감상평

- 이번 진도는 대체적으로 쑥쑥 읽기 좋았습니다.
- 간혹 회로나 수학 등 순수학문 같은 개념으로 파고들면 머리가 지끈했지만요.
- 이미 알고 있는 개념과 함께 조합해 재밌고 이해하기 좋았어요.

<br/>

## 4.1장: 이 작은 장난감을 CPU라고 부른다

- 초반에 회로 이야기를 다룰 때는 오엥... ㅇ_ㅇ? 했다가 점차 이해가 잘 되는 흐름이었어요.
- 특히 **명령어 집합이 소프트웨어와 하드웨어가 서로 통신하는 인터페이스** 라는 점 등등
- **CPU(프로세서)의 동작 원리가 프로그래밍 언어의 동작 원리와 맞닿아있다**는 설명이 잘 와닿았습니다.
- Claude의 도움으로 4.1.4장의 🧮 [간단한 가산기(adder)](https://codepen.io/ella-yschoi/pen/mdZVZEN)를 만들어 보았습니다. 재미삼아 해보세요! 
- 책에서 논리합, 논리곱 등등 몇 가지 연산을 섞어 설명해 헷갈렸는데, 논리연산에 관해서는 [나무위키](https://namu.wiki/w/%EB%85%BC%EB%A6%AC%20%EC%97%B0%EC%82%B0)가 잘 요약해주네요.
- 4.1.6장의 회로가 전원이 끊기면 저장된 정보가 사라진다길래, 앞서 3장에서도 SSD와의 차이를 언급하며 유사한 설명을 했던 기억이 났어요.
- 찾아보니 레지스터와 RAM은 주로 [플립플롭(Flip-flop)](https://ko.wikipedia.org/wiki/%ED%94%8C%EB%A6%BD%ED%94%8C%EB%A1%AD) 회로를 기반으로 구성되며, 뒤에 나오는 클럭 신호가 필요하다고 합니다. 깨알지식! 🤓
- 레지스터와 RAM 모두 휘발성이지만, 그 중에서 레지스터와 SRAM은 플립플롭 기반이고, DRAM은 커패시터 기반이라고 하네요.

<br/>

## 4.2장: CPU는 유휴 상태일 때 무엇을 할까?

- 읽으면서 든 생각: 아니 CPU가 유휴 상태면 좀 쉬게 해줘라! 왜 유휴 프로세스라는 걸 또 만드는거야
- 간혹 쉬어도 되는데 쉬질 못하고 뭐라도 해야하는 제 모습과 흡사하다는 생각이 들었습니다 🫠
- 하지만 **이 유휴 프로세스의 역할은 CPU 내부를 절전 상태로 전환해 전력 소비를 줄이기 위한,**
- 나중에 더 좋은 성능으로 일하기 위해 잘 쉬어 두려고 하는 프로세스 같다는 생각을 했어요.
- 우리도 잘 일하기 위해 잘 쉽시다 (?) ... 타이머 인터럽트가 있기 전에.

<br/>

## 4.3장: CPU는 숫자를 어떻게 인식할까?

- 갑자기 수학책이 된 것 같았던 4.3장... 😵‍💫
- 어쨌든 요점은 **CPU가 숫자를 잘 인식하게 표현할 수 있을지 고민하면서, 이렇게 다양한 방법을 시도했다~**인 것 같습니다.
- 그리고 CPU가 하는 일과 프로그래머가 하는 일은 생각보다 간극이 크구나, 느꼈습니다.
- 이 간극을 크게 만들어 준 여러 장치와 연구들에 감사하다.. 덕분에 편하게 개발할 수 있다.. (지난 시간에 이어 또 감사🙏🏻)
- 1의 보수, 2의 보수에 관해서는 책 설명으로는 이해가 조금 어려워 시각화 해놓은 이미지를 가져와 보았습니다.

> 출처: https://lifelectronics.tistory.com/69

<img width="919" alt="image" src="https://github.com/user-attachments/assets/be7de94e-8c9d-43af-8b73-b686f83289f8">

<br/>

## 4.4장: CPU가 if문을 만났을 때

- 코드를 쓰면서 '**CPU를 더 잘 활용하는 코드**'에 대해서는 염두하지 않았던 것 같은데 이 장을 읽으면서 또 다른 관점을 +1 획득했습니다.
- 알고리즘 문제를 풀면서 특히 배열을 많이 사용하는데 그때 성능 어쩌고 하는 이유를 더 깊게 알 것 같은 느낌
- 어쨌든 **배열 요소가 이미 정렬되어 있으냐, 무작위로 배치되어 있느냐**가 성능에 꽤 큰 영향을 미친다는 것을 깨달았습니다.
- 마치 현실에서도 딱딱 정리되어 어디에 뭐가 있는지 알면 찾을 때 에너지가 덜 쓰이는 것처럼요 🗂️

<br/>

## 4.5장: CPU 코어 수와 스레드 수 사이의 관계는 무엇일까?

- 4.5장 내용은 뒤의 **CPU 진화론을 다루기 위한 티저**에 가깝군요.
- 더불어 CPU, 운영체제, 스레드, 프로세스를 공부하며 헷갈릴만한 것들을 잘 짚어준 것 같아요. 예를 들면 이런 것들.
  - CPU 코어 수와 스레드 수 사이에는 필연 관계가 없다,
  - 단일 코어 시스템에서는 스레드 여러 개가 번갈아 실행되긴 하지만 진정한 병렬 처리라고는 할 수 없다,
  - 그렇다고 단일 코어에서는 다중 스레드가 의미가 없진 않다,
  - 다중 프로세스 프로그래밍은 번거롭고 프로세스 간 전환에 비용이 들지만, 스레드 덕분에 다중 코어 리소스를 효율적으로 사용 가능하다,
  - 스레드 수는 코어 수와 일정한 선형 관계를 유지해야 한다, 그렇다고 스레드가 많다고 좋은 것은 아니다,
- 어쨌든 **'적당히' 라는 것은 프로그램과 상황마다 다르니 위의 점들을 인지**하고 있는 것이 좋을 것 같습니다.

<br/>

**Reply by yunseorim1116** [reply]
2의 보수 헷갈렸는데... 댑악... 연수님 감사합니다 !!!!!!!!

**Reply by ddongule** [reply]
당신을 정리왕에 임명합니다

**Reply by ddongule** [reply]
쿨럭 쿨럭쿨럭 기침신호 (ㅈㅅ합니다)

**Comment by yunseorim1116** [comment]
한동안 v8에 빠져서 한 2월쯤이었나.. 설날 연휴에 이틀동안 혼공컴 책을 급급하게 읽었던적이 있었다. 그때는 ALU, 레지스터, 클럭 등 용어를 단순암기하기에 바빠 시간이 지나고 거의 잊어버렸다. 이 책에서는 논리곱 게이트, 논리합 게이트, 논리부정 게이트와→ 가산기→ ALU(산술 논리 장치) 를 차례로 설명을 해주기 때문에 무작정 외우게 시키는게 아니라, 자연스럽게 흐름을 따라가게 만드는 것 같았다. 그래서 더 기억에 오래오래 남을듯.

<img width="340" alt="image" src="https://github.com/user-attachments/assets/22c0e049-682f-4d10-aa7d-0adc55cb96ae">


추가적으로 v8 공부할때 사수님이 '태깅' 이라는 기법을 설명해주시면서 추가적으로 이진수 숫자에서 음수를 표현하는 방법을 가르쳐주셨는데 이 책에도 그 내용이 나와서 너무 반가왔다. 그때는 몰라서 당당하게 모른다고 말씀드렸는데 .. 이제는 당당하게 말할수 있길......... ^^

**Reply by ella-yschoi** [reply]
맞아요! 이 책의 매력은 평소에는 어렵고 머릿속에 그려지지 않는 개념을 -> 자연스럽게 흐름을 따라가게 만든다는 점에 공감합니다!
ㅋㅋㅋㅋㅋㅋㅋ 회사에서 서림님 닉네임 Tree 인가요? 귀욥.. 🌳🌳 설마 한자 '수풀림[林]'이라서 그런 건 아니죠....?

**Reply by yunseorim1116** [reply]
사실 제 이름의 림은 琳 이구용ㅎㅁㅎ(tmi) 사주에 나무가 부족해서 이전 회사부터 꾸준히 Tree라는 닉네임을 쓰고있답니다,,,ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ닉네임은 트리인데 트리구조를 잘 몰라서 공부해야겠다고 생각만 하고있어요 허허허(?

**Reply by ella-yschoi** [reply]
오오 '옥 림[琳]'도 정말 좋은 의미네요!
사주에 나무가 부족해서 닉네임을 Tree로 쓰신다니 넘나 신박한,, 👍👍

**Reply by hanana1253** [reply]
모릅니다~~ ㅋㅋㅋ 넘귀욤

**Comment by sunhpark42** [comment]
저자분.. 컴퓨터 구조를 좋아하시는 걸까. 묘하게 글에서 신남이 느껴졌다.
CPU로 할 수 있는 논리연산 자체는 아주 간단하지만 이걸 어떻게든 발전시킨 사람들이 대단하다. 이래서 수학잘하면 컴퓨터도 이해를 잘하는걸까.. 뭐가 됐던 나는 아닌것 같다.
여전히 CPU 구조는 노잼

덧) CPU, GPU 차이를 설명하는 간단한 영상 ([https://youtu.be/1BAZf3PsjWA?si=mOvcx9sOWT8xTv79](https://youtu.be/1BAZf3PsjWA?si=mOvcx9sOWT8xTv79))

**Comment by sa02045** [comment]
컴퓨터는 마법이 아닌라는걸 다시 깨닫게된 챕터였습니다

- 논리함수를 컴퓨터가 어떻게 표현할 수 있을까?
   - 트랜지스터 -> 논리곱, 논리합,논리부정 게이트 -> 논리 완전성! -> 모든 논리를 구현할 수 있다
   - 논리게이트로 기억(메모리)도 할 수 있다
   - 더하기 연산, 빼기 연산등등의 연산 -> ALU가 처리
- CPU가 숫자를 인식하는 방법
   - 간단 버전(최상위 비트가 정수부호결정하는)부터 똑똑하고 효율적인 버전까지(2의보수)
- 배열을 정리한 뒤 처리해야하는 이유!
  - CPU가 추측을 잘 할 수 있도록
- 예전에 얘기나누었던 CPU 코어, 스레드, 프로세스, 동시성, 병렬성의 관계

**Comment by hanana1253** [comment]
4-1~4-5는 쉬어가는 챕터라 생각될 정도로 이해하기에 어렵지 않았다. 그래서 정리는 따로 안함. 

다양한 예시를 들어 설명해줘서 이해가 쉬웠던 듯. 모처럼 친절한 책을 만나니 과거의 SICP가 얼마나 모진 책이었는지 새삼 깨닫는다. 

4-1은 중간에 포기.. 넘 복잡해

> p.300 프로그램에 무한 순환이 있더라도 운영 체제는 여전히 타이머 인터럽트를 통해 프로세스의 스케줄링을 제어할 수 있으며 ... 정말 설계 방식이 우아하지 않나요? 

ㅇㅈ... 



작업관리자 화면을 이 생에서 이해할 수 있을 거라고 생각 못했는데 뭔지 알게되어 신기하다. 

1의 보수나 2의 보수라는 체계에 대해 어떻게 생각해냈을까? 신기... 

Cpu자체가 인간 두뇌에 존재하는 개념을 이해할 수 없다는 게 뭔가 안심이 된다. AI가 인류를 넘어서진 못하겠...지?

if 분기예측...! cpu도 때려맞춰서 하는 것이라니 의외였다

> p.301 컴퓨터는 손가락이 하나하서 2진법입니...... 하하! 농담입니다. 

부장님이신줄...

**Reply by hanana1253** [reply]
코드(code)라는 책을 다시 읽어봐야겠다...

**Reply by hanana1253** [reply]
Fallacy unit? 
바코드에서의 맨앞/맨뒤 바

**Reply by ddongule** [reply]
![image](https://github.com/user-attachments/assets/fc3747f0-ac86-4cef-a219-b307ea813c59)

**Comment by iborymagic** [comment]
- 항상 요런 CS 책을 읽어보면 저자들이 트랜지스터 ~ 회로 파트를 빌드업하면서 CPU의 위대함을 찬양하는 것 같은 느낌이 듭니다.. 허허(몇 개 안읽어봄)
  - 스위치에서 CPU까지의 발전 과정을 읽다보면 대단하다는 생각이 들지만, 개인적으로는 가장 재미없는 파트인 것 같아서 맨날 대충 읽게 됩니다..
- 분기 예측 성공 여부에 따른 성능 차이는 꽤나 흥미로웠읍니다.. 
  - 어찌보면 당연하겠지만, 메모리뿐만 아니라 CPU를 고려한 코드라는 게 뭔가 신선하게 다가왔던 것 같읍니다.. 맨날 이 파트를 대충 읽고 넘어가서 CPU와 성능이 연관될 수 있다는 생각을 안했나봅니다..
  - 나중에 복잡한 컴퓨팅이 필요한 기능을 만들 때, perf 명령어를 한 번쯤 활용해봐도 되겠다는 생각이 들었읍니다... 
  - 물론 드라마틱한 개선이 이뤄질지는 모르겠지만.. 그냥 저런거 캡쳐해서 수치가 달라진 걸 보여주면 생색내기 좋으니까요 허허
- 분기 예측, 정렬의 중요성 등을 읽으면서, 뭔가 요즘 컴퓨터들이 성능이 너무 좋아져서.. 요런 사소한 코드적인 성능 부분은 너무 고려를 하지 않은 것은 아닌가 하는 괜한 반성이 들었읍니다..
- 스레드 수가 너무 많아져도 전환 비용이 높아져 성능에 악영향을 미칠 수 있구나.. 하는 것을 깨달았읍니다.. 진리의 과유불급..

**Comment by chloe-codes1** [comment]
## 감상평

- 트랜지스터의 논리 연산을 이렇게 이해하기 쉽게 설명한 글이 있었나 싶을 정도로 술술 읽혔다.
  - 앨런 튜링 대단해요…
- 제조업의 파이프라인 기술을 CPU의 효율성과 연관지어 설명한 부분이 내용을 이해하기 쉽게 해줬다!

**Comment by thgr8ganzi** [comment]
* CPU 가 숫자를 인식할때 원래는 개념이 짬뽕되어 있었는데 이 책을 읽고 개념이 정리 되었다.
* if 문을 만났을때 CPU가 분기 예측을 한다는게 신기하다. 현대 CPU는 얼마나 성능이 좋은것인지 가늠이 안된다.
* 이번 챕터는 실제 프로그래밍을 하면서 쓸지 의문이 든다.

**Comment by ddongule** [comment]
## 감상평
- 회로부터 진짜 컴퓨터를 뜯어보는 느낌이었다. 물론 제대로 이해를 했는지는 .. ? .. ㅎㅎ.. 수학 쪽을 내가 너무 이해하려고 해서 그런가.. 흠..^^ !!!!! (애써 이유를 찾는 중) 감상평을 쭉 살피는데 다들 쉬웠다고 해서 내 뇌가 술을 많이먹어서 쪼그라들었구나! 라고 생각했다
- 사실 그래도 다른 책들보다는 확연히 쉽게 설명이 되어있는 것 같다. 으하학
- 이미테이션 게임을 재밌게 봤는데 앨런튜링이 나와서 반가웠다. 하이! 암호만 잘하는게 아니었구나? 
- 컴퓨터가 느리면 작업 관리자 들어가서 괜히 많이 잡아먹고있는 애들 보고 kill 하고 그랬는데 idle process가 몬지 나도 궁금했지만 찾아보지 않았었는데 또 이렇게 책을 통해 알게되어 재미났다. 
- CPU를 메가팩토리와 파이프라인 기술에 비유한거 재밌었다. 근데 정렬 되어있을때 안되어있을 때 속도가 아니라 정확도가 많이 차이가 난다는게 신기했다. 알아서 정렬 잘 할 줄 알았는데? ㅎ;; 속도..는 똑같으려나? 다시 찾아봤는데 속도 얘기는 내가 못찾은건지 딱히 없어서 궁금했다. 
- 다 읽고 나서는 내가 제대로 이해한건지 ?? 이런 느낌이었는데 막상 이렇게 감상평 정리할라다보니 엄청 어렵진 않았던 것 같기도 하고 ^^? (?) 첨에 이번 장에서 작가 너는 나를 이해시키는데 실패햇지 라고 꼽주려다가 다시 보니 나쁘지 않았던 것 같기도 하다 ㅎㅎㅎㅎㅎㅎㅎㅎㅎㅎ 여자의 마음은 갈대야~

**Reply by ddongule** [reply]
갑자기 든 의문: CPU는 유휴 상태이면 문제가 생기나요? 허허헛
